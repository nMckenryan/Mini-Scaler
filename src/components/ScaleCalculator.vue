<template>
  <v-container width="200" class="mx-auto mt-5">
    <v-row class="text-center">
      <v-col class="mb-5 mx-auto" cols="6">
        <v-form>
          <v-layout row wrap>
            <!-- CENTIMETRES -->
            <v-flex s2>
              <v-text-field
                label="Character Height (Centimetres)"
                prepend-icon="mdi-tape-measure"
                :value="realCm"
                type="number"
                suffix="cm"
                @input="updateCm"
              ></v-text-field>

              <!-- Inches -->
              <v-text-field
                label="Character Height (Inches)"
                prepend-icon="mdi-tape-measure"
                :value="realI"
                type="number"
                suffix="in"
                @input="updateIn"
              ></v-text-field>
            </v-flex>

            <v-spacer></v-spacer>

            <!-- METRES -->
            <v-flex s2>
              <v-text-field
                label="Character Height (Metres)"
                prepend-icon="mdi-tape-measure"
                :value="realM"
                type="number"
                suffix="m"
                @input="updateM"
              ></v-text-field>

              <v-spacer></v-spacer>

              <!-- Feet -->
              <v-text-field
                label="Character Height (Ft)"
                prepend-icon="mdi-tape-measure"
                :value="realFt"
                type="number"
                suffix="ft"
                @input="updateFt"
              ></v-text-field>
            </v-flex>
          </v-layout>

          <v-spacer></v-spacer>

          <!-- Scale Selection -->
          <v-btn-toggle
            v-model="scaleType"
            tile
            color="green accent-4"
            group
            mandatory
          >
            <v-btn value="15">
              15mm
            </v-btn>

            <v-btn value="28">
              28mm
            </v-btn>

            <v-btn value="32">
              32mm
            </v-btn>

            <v-btn value="54">
              54mm
            </v-btn>
          </v-btn-toggle>

          <v-text-field
            label="Printed Mini Size"
            v-model.lazy="scaleToMini"
            prepend-icon="mdi-scale-balance"
            type="number"
            suffix="mm"
          ></v-text-field>
        </v-form>
      </v-col>
    </v-row>
  </v-container>
</template>

<script>
export default {
  props: ["hPreset"],
  name: "ScaleCalculator",

  data: () => ({
    realCm: 0,
    realI: 0,
    realFt: 0,
    realM: 0,
    scaleType: "28",
    scaleHeight: 0,
  }),

  watch: {
    hPreset: function() {
      //Handles Top menu preset changes.
      this.updateCm(this.hPreset);
    },
  },

  computed: {
    scaleToMini: function() {
      let equation = 1;
      switch (this.scaleType) {
        case "15":
          equation = 12.2;
          break;
        case "28":
          equation = 6.5;
          break;
        case "32":
          equation = 5.7;
          break;
        case "54":
          equation = 3.39;
          break;
      }
      return (this.realCm / equation).toFixed(2);
    },
  },

  methods: {
    // TODO: Test and verify conversions are correct
    updateCm(value) {
      this.realI = value * 0.39;
      this.realFt = value * 0.032;
      this.realCm = value;
      this.realM = value * 0.01;
    },

    //TODO: Sort out ft/inch display
    updateIn(value) {
      this.realI = value;
      this.realFt = value * 12;
      this.realCm = value * 2.54;
      this.realM = value * 0.0254;
    },

    updateM(value) {
      this.realI = value * 39.37;
      this.realFt = value * 3.28;
      this.realCm = value * 100;
      this.realM = value;
    },

    updateFt(value) {
      this.realI = value * 12;
      this.realFt = value;
      this.realCm = value * 30.48;
      this.realM = value * 0.3;
    },
  },
};
</script>
