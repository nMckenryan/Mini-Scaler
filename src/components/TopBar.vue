<template>
  <div class="mx-auto overflow-hidden" height="400">
    <v-app-bar app color="light-green darken-4" dark>
      <v-app-bar-nav-icon @click="drawer = true"></v-app-bar-nav-icon>
      <!-- TITLE -->
      <v-img
        alt="MiniscalerLogo"
        class="shrink mr-2"
        contain
        src="..\assets\scalerLogoWhite.png"
        transition="scale-transition"
        width="45"
      />

      <v-toolbar-title>Mini Scaler</v-toolbar-title>
    </v-app-bar>

    <v-navigation-drawer v-model="drawer" absolute temporary>
      <!-- TODO: Close Menu on click -->

      <v-list>
        <v-list-group
          v-for="item in items"
          :key="item.title"
          v-model="item.active"
          :prepend-icon="item.action"
          no-action
          color="primary"
        >
          <template v-slot:activator>
            <v-list-item-content>
              <v-list-item-title v-text="item.title"></v-list-item-title>
            </v-list-item-content>
          </template>

          <v-list-item two-line v-for="child in item.items" :key="child.title">
            <a @click="setPreset(child.height)">
              <v-list-item-content>
                <v-list-item-title v-text="child.name"></v-list-item-title>
                <v-list-item-subtitle
                  v-text="child.height + 'cm'"
                ></v-list-item-subtitle> </v-list-item-content
            ></a>
          </v-list-item>
        </v-list-group>
      </v-list>

      <!-- 
      <v-list nav dense color="light-green lighten-4">
        <v-list-group
          v-for="item in items"
          :key="item.title"
          v-model="item.active"
          :prepend-icon="item.action"
          no-action
        >
          <v-list-item disabled bolder backgroundcolor="red">
            40k References
          </v-list-item>

          <v-list-item v-for="(item, i) in Wh40Heights" :key="i">
            <v-list-item-name>
              <a @click="setPreset(item.height)">
                {{ item.name }} {{ item.height }} cm</a
              ></v-list-item-name
            >
          </v-list-item>

          <v-list-item>
            DnD References
          </v-list-item>
          <v-list-item v-for="(item, i) in dndHeights" :key="i">
            <v-list-item-name>
              <a @click="setPreset(item.height)">
                {{ item.name }} {{ item.height }} cm</a
              >
            </v-list-item-name>
          </v-list-item>
        </v-list-group>
      </v-list> -->
    </v-navigation-drawer>

    <Scale-Calculator v-bind:hPreset="menuSelect" />
  </div>
</template>

<script>
import ScaleCalculator from "./ScaleCalculator.vue";

export default {
  name: "Topbar",
  components: {
    ScaleCalculator,
  },

  data: () => ({
    drawer: false,
    items: [
      {
        action: "mdi-sword-cross",
        title: "DnD Heights",
        active: true,
        items: [
          { name: "Human", height: 170 },
          { name: "Half Elf", height: 167 },
          { name: "Elf", height: 165 },
          { name: "Dwarf", height: 125 },
          { name: "Halfling", height: 91 },
          { name: "Gnome", height: 101 },
          { name: "Half-Orc", height: 175 },
          { name: "Dragonborn", height: 190 },
          { name: "Goliath", height: 228 },
        ],
      },
      {
        action: "mdi-hammer",
        title: "40K Heights",
        items: [
          { name: "Imperial Guardsman", height: 182 },
          { name: "Space Marine (Armoured)", height: 228 },
          { name: "Ork ", height: 198 },
          { name: "Ork Warboss ", height: 300 },
          { name: "Genestealer ", height: 189 },
          { name: "Eldar/Drukhari ", height: 189 },
          { name: "Necron ", height: 243 },
          { name: "Tau", height: 167 },
          { name: "Tyranid Warrior", height: 274 },
        ],
      },
    ],
  }),

  methods: {
    setPreset(item) {
      this.menuSelect = item;
    },
  },
};
</script>
