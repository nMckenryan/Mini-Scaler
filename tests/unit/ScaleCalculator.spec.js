import ScaleCalculator from "@/components/ScaleCalculator.vue";
import TopBar from "@/components/TopBar.vue";
import { mount } from "@vue/test-utils";

// test("mount Calc", () => {
//   const wrapper = mount(ScaleCalculator);
//   expect(wrapper.html()).toMatchSnapshot();
// });

test("mount TopBar", () => {
  const wrapper = mount(TopBar);
  expect(wrapper.html()).toMatchSnapshot();
});

// The component to test
const MessageComponent = {
  template: "<p>{{ msg }}</p>",
  props: ["msg"],
};

test("check Props", () => {
  // mount() returns a wrapped Vue component we can interact with
  const wrapper = mount(MessageComponent, {
    propsData: {
      msg: "Hello world",
    },
  });

  // Assert the rendered text of the component
  expect(wrapper.text()).toContain("Hello world");
});
