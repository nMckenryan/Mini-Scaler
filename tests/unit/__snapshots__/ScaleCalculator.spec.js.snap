// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mount TopBar 1`] = `
<div>
  <v-app-bar app="" color="light-blue darken-4" dark="">
    <div class="d-flex align-center">
      <v-img alt="MiniscalerLogo" contain="" src="..\\assets\\scalerLogoWhite.png" transition="scale-transition" width="45" class="shrink mr-2"></v-img>
      <h1 class="display-1 font-weight-bold my-auto">
        Mini Scaler
      </h1>
    </div>
    <v-spacer></v-spacer>
    <v-menu transition="slide-y-transition" bottom="">
      <v-list>
        <v-list-item>
          <v-list-item-name><a>
              Imperial Guardsman 182 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Space Marine (Armoured) 228 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Ork 198 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Ork Warboss 300 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Genestealer 189 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Eldar/Drukhari 189 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Necron 243 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Tau 167 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Tyranid Warrior 274 cm</a></v-list-item-name>
        </v-list-item>
      </v-list>
    </v-menu>
    <v-menu transition="slide-y-transition" bottom="">
      <v-list>
        <v-list-item>
          <v-list-item-name><a>
              Human 170 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Half Elf 167 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Elf 165 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Dwarf 125 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Halfling 91 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Gnome 101 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Half-Orc 175 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Dragonborn 190 cm</a></v-list-item-name>
        </v-list-item>
        <v-list-item>
          <v-list-item-name><a>
              Goliath 228 cm</a></v-list-item-name>
        </v-list-item>
      </v-list>
    </v-menu>
  </v-app-bar>
  <v-container width="200" class="mx-auto mt-5">
    <v-row class="text-center">
      <v-col cols="8" class="mb-5 mx-auto">
        <v-form>
          <v-layout row="" wrap="">
            <v-flex s2="">
              <v-text-field label="Character Height (Centimetres)" prepend-icon="mdi-tape-measure" value="0" type="number" suffix="cm"></v-text-field>
            </v-flex>
            <v-spacer></v-spacer>
            <v-flex s2="">
              <v-text-field label="Character Height (Inches)" prepend-icon="mdi-tape-measure" value="0" type="number" suffix="in"></v-text-field>
            </v-flex>
          </v-layout>
          <v-layout>
            <v-flex s2="">
              <v-text-field label="Character Height (Metres)" prepend-icon="mdi-tape-measure" value="0" type="number" suffix="m"></v-text-field>
            </v-flex>
            <v-spacer></v-spacer>
            <v-flex s2="">
              <v-text-field label="Character Height (Ft)" prepend-icon="mdi-tape-measure" value="0" type="number" suffix="ft"></v-text-field>
            </v-flex>
          </v-layout>
          <v-spacer></v-spacer>
          <v-btn-toggle tile="" color="deep-blue accent-3" group="" required="">
            <v-btn value="28">
              28mm
            </v-btn>
            <v-btn value="32">
              32mm
            </v-btn>
            <v-btn value="32">
              54mm
            </v-btn>
          </v-btn-toggle>
          <v-text-field label="Printed Mini Size" prepend-icon="mdi-scale-balance" type="number" suffix="mm" readonly=""></v-text-field>
        </v-form>
      </v-col>
    </v-row>
  </v-container>
</div>
`;
